pos_accounting/src/pool/storage/mod.rs

```diff
@@ -20,7 +28,169 @@ impl<'a, S> PoSAccountingDBMut<'a, S> {
+            |s, id| s.del_pool_balance(id),
+        )?;
+
+        self.merge_balances_generic(
+            other.delegation_balances.consume().into_iter(),
+            |s, id| s.get_delegation_balance(id),
+            |s, id, amount| s.set_delegation_balance(id, amount),
+            |s, id| s.del_delegation_balance(id),
+        )?;
+
+        self.merge_balances_generic(
+            other.pool_delegation_shares.consume().into_iter(),
+            |s, (pool_id, delegation_id)| s.get_pool_delegation_share(pool_id, delegation_id),
+            |s, (pool_id, delegation_id), amount| {
+                s.set_pool_delegation_share(pool_id, delegation_id, amount)
+            },
+            |s, (pool_id, delegation_id)| s.del_pool_delegation_share(pool_id, delegation_id),
+        )?;
+
+        Ok(DataMergeUndo {
+            pool_data_undo,
+            delegation_data_undo,
+        })
+    }
+
+    pub fn undo_merge_with_delta(
+        &mut self,
+        other: PoSAccountingDeltaData,
+        undo: DataMergeUndo,
+    ) -> Result<(), Error> {
+        undo.pool_data_undo
+            .into_iter()
+            .try_for_each(|(key, undo_data)| match undo_data {
+                Some(data) => self.store.set_pool_data(key, data),
+                None => self.store.del_pool_data(key),
+            })?;
+
+        undo.delegation_data_undo
+            .into_iter()
+            .try_for_each(|(key, undo_data)| match undo_data {
+                Some(data) => self.store.set_delegation_data(key, data),
+                None => self.store.del_delegation_data(key),
+            })?;
+
+        self.merge_balances_generic(
+            other.pool_balances.consume().into_iter().map(|(k, v)| (k, v.neg().unwrap())),
+            |s, id| s.get_pool_balance(id),
+            |s, id, amount| s.set_pool_balance(id, amount),
+            |s, id| s.del_pool_balance(id),
+        )?;
+
+        self.merge_balances_generic(
+            other
+                .delegation_balances
+                .consume()
+                .into_iter()
+                .map(|(k, v)| (k, v.neg().unwrap())),
+            |s, id| s.get_delegation_balance(id),
+            |s, id, amount| s.set_delegation_balance(id, amount),
+            |s, id| s.del_delegation_balance(id),
+        )?;
+
+        self.merge_balances_generic(
+            other
+                .pool_delegation_shares
+                .consume()
+                .into_iter()
+                .map(|(k, v)| (k, v.neg().unwrap())),
+            |s, (pool_id, delegation_id)| s.get_pool_delegation_share(pool_id, delegation_id),
+            |s, (pool_id, delegation_id), amount| {
+                s.set_pool_delegation_share(pool_id, delegation_id, amount)
+            },
+            |s, (pool_id, delegation_id)| s.del_pool_delegation_share(pool_id, delegation_id),
+        )?;
+
+        Ok(())
+    }
+
+    fn merge_balances_generic<Iter, K: Ord + Copy, Getter, Setter, Deleter>(
```

Comment:
Please throw in TODOs to test merge_balances_generic and merge_data_generic independenly, since they're now perfectly abstracted.